_posts/2021-05-03-deploying-kafka-with-lets-encrypt-certificates.md

```diff
@@ -0,0 +1,487 @@
+
+In addition to that, the `brokerCertChainAndKey` section specifies the listener certificate used for this listener.
+We just point it to the Kubernetes Secret created by cert-manager that contains our signed certificate.
+
+The `Kafka` custom resource I've shown above is just an example. It doesn't specify authentication, metrics, and other configuration.
+But it should help you to get started.
+
+## Testing the certificate with OpenSSL
+
+Once the Kafka cluster is deployed, we can use OpenSSL to check that it is using the right certificate.
+We use the `s_client` feature and check that it uses the right certificate signed by Let's Encrypt:
+
+```
+openssl s_client -connect bootstrap.example.com:443 -servername bootstrap.example.com
+CONNECTED(00000005)
+depth=2 O = Digital Signature Trust Co., CN = DST Root CA X3
+verify return:1
+depth=1 C = US, O = Let's Encrypt, CN = R3
+verify return:1
+depth=0 CN = bootstrap.example.com
+verify return:1
+---
+Certificate chain
+ 0 s:/CN=bootstrap.example.com
+   i:/C=US/O=Let's Encrypt/CN=R3
+ 1 s:/C=US/O=Let's Encrypt/CN=R3
+   i:/O=Digital Signature Trust Co./CN=DST Root CA X3
+---
+Server certificate
+-----BEGIN CERTIFICATE-----
+...
+-----END CERTIFICATE-----
+subject=/CN=bootstrap.example.com
+issuer=/C=US/O=Let's Encrypt/CN=R3
+---
+No client certificate CA names sent
+Server Temp Key: ECDH, X25519, 253 bits
+---
+SSL handshake has read 2970 bytes and written 317 bytes
+---
+New, TLSv1/SSLv3, Cipher is ECDHE-RSA-AES256-GCM-SHA384
+Server public key is 2048 bit
+Secure Renegotiation IS supported
+Compression: NONE
+Expansion: NONE
+No ALPN negotiated
+SSL-Session:
+    Protocol  : TLSv1.2
+    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
+    Session-ID: F2CE7100EF5F4E077C69EA3CAE6307F6F90848933FAAEF2F34F1F79C6A35590A
+    Session-ID-ctx:
+    Master-Key: 46A6C5A04A034E9350D73AC037E52D4CACD49CF6F33103B929372CB7F44816939A5473C9E67DD6AFE03D221ADA6412BB
+    Start Time: 1619999991
+    Timeout   : 7200 (sec)
+    Verify return code: 0 (ok)
+---
+...
+```
+
+As you can see from the following part of the output:
+
+```
+Certificate chain
+ 0 s:/CN=bootstrap.example.com
+   i:/C=US/O=Let's Encrypt/CN=R3
+ 1 s:/C=US/O=Let's Encrypt/CN=R3
+   i:/O=Digital Signature Trust Co./CN=DST Root CA X3
+```
+
+the server certificate used by `bootstrap.example.com` is signed by Let's Encrypt.
+
+## Sending or receiving messages
+
+We can now also test that we can produce and consume messages from the Kafka cluster without configuring a certificate in our clients.
+When using the console consumer or producer from Kafka, we need to specify the `--bootstrap-server` option and the `--topic` option.
+And in addition to that we will also enable TLS encryption with `--consumer-property security.protocol=SSL`.
+We do not need to specify any truststore.
+We will use the default CA certificates bundled with Java.
```

Comment:
I would highlight that that CA certificates default truststore with Java contains the Let's Encrypt CA certificate and for this reason it works oob.

Reply:
I'm not sure I want to get into this level of details ... does it really trust Let's Encrypt (ISRG Root X1)? Or does it trust IdenTrust DST Root CA X3? It will probably depend on version of Java etc.

Reply:
The OpenSSL command example in the blog post for example shows that up-to-date MacOS trusts it through IdenTrust ... so I would simply avoid going into more details because it is more complicated.